### PR TITLE
implement ECS mapping migration tooling

### PR DESCRIPTION
**Pull request by efd6**
Please take a look.

### CODE REVIEW

File: ecs-update/testdata/my_package/data_stream/item_usages/fields/ecs.yml [hunk 1]
```diff
@@ -44,5 +40,3 @@
   name: source.geo.region_name
 - external: ecs
   name: source.ip
-- external: ecs
-  name: tags
```
Comment (efd6) [inline ecs-update/testdata/my_package/data_stream/item_usages/fields/ecs.yml:48]: Moved this to base-fields.yml to exercise partial removal.

File: ecs-update/testdata/my_package/data_stream/item_usages/fields/ecs.yml [hunk 2]
```diff
@@ -20,10 +20,6 @@
   name: user.full_name
 - external: ecs
   name: user.email
-- external: ecs
-  name: os.name
-- external: ecs
-  name: os.version
```
Comment (efd6) [inline ecs-update/testdata/my_package/data_stream/item_usages/fields/ecs.yml:26]: Removed these because they were not being recognised as ECS fields.

File: ecs-update/ecs-update.go [hunk 3]
```diff
@@ -856,6 +926,30 @@ func (e *packageEditor) modifyManifest() error {
 		e.result.Manifest.FormatVersionNew = e.config.Manifest.FormatVersion
 	}
 
+	if e.config.Manifest.KibanaVersion != "" && e.config.Manifest.KibanaVersion != e.pkg.Manifest.Conditions.Kibana.Version {
+		var haveStack, wantStack versionConstraint
+		err = haveStack.Set(e.pkg.Manifest.Conditions.Kibana.Version)
```
Comment (andrewkroh) [inline ecs-update/ecs-update.go:931]: Values like this https://github.com/elastic/integrations/blob/0cd817a50cfd03e49b1dbd4d21b8daa91b0e48a3/packages/fireeye/manifest.yml#L12 fail to parse.

```json
  {
    "package": "fireeye",
    "changed": false,
    "failed": true,
    "error": "failed to parse existing version: strconv.ParseInt: parsing \"0 || ^8.0.0\": invalid syntax"
  }
```

The package-registry uses https://pkg.go.dev/github.com/Masterminds/semver/v3#Constraints to parse this. We should either be able to accept the same format or use that library. If you use that library I'm not sure you can implement an equivalent `haveStack.LessThan(want)` check to determine if it needs raised.

<details>

<summary>Breadcrumbs to find the version constraint format used in packages</summary>

- https://github.com/elastic/package-spec/blob/a895b110b5a86b6ae5410c8bd3834539e27a1b92/spec/integration/manifest.spec.yml#L140
- https://github.com/elastic/package-spec/blob/a895b110b5a86b6ae5410c8bd3834539e27a1b92/code/go/internal/validator/semantic/validate_minimum_kibana_version.go#L151-L164
- https://github.com/elastic/package-registry/blob/1c6fed752fa5c55ffed40b50805bffac3fddd21f/packages/package.go#L121
- https://pkg.go.dev/github.com/Masterminds/semver/v3#Constraints

</details>

File: ecs-update/ecs-update.go [hunk 4]
```diff
@@ -804,15 +872,27 @@ func (e *packageEditor) modifyBuildManifest() error {
...
 
-	e.result.BuildManifest.Changed = true
-	e.result.BuildManifest.ECSReferenceOld = e.pkg.Build.Dependencies.ECS.Reference
-	e.result.BuildManifest.ECSReferenceNew = e.config.BuildManifest.ECSReference
+	if e.config.BuildManifest.ECSImportMappings != nil && !*e.config.BuildManifest.ECSImportMappings {
+		err = yamlDeleteNodeFromMap(f, "$.dependencies.ecs.import_mappings")
+		if err != nil {
+			return err
+		}
+		e.result.BuildManifest.Changed = true
```
Comment (andrewkroh) [inline ecs-update/ecs-update.go:892]: I would rename `Changed` to `ECSReferenceChanged`. And then create a separate `ECSImportMappingsChanged`. This way the changelog and commit messages can reflect specifically what changed.

Comment (andrewkroh) [inline ecs-update/ecs-update.go:892]: I think this is being set to true even when the no change occurs which makes it add a changelog entry and do a commit.

Running this reproduces the issue, for example.

`ecs-update -drop-import-mappings packages/zoom`

File: ecs-update/ecs-update.go [hunk 5]
```diff
@@ -1798,13 +1937,85 @@ func fieldsYMLUseExternalECS(f *ast.File, ff fleetpkg.FieldsFile) (changed bool,
...
+	}
+
+	for _, field := range flatFields {
+		// Does the field exist in ECS?
+		ecsField, _ := ecs.Lookup(field.Name, "")
+		if ecsField == nil {
+			continue
+		}
+
+		// The type must be the same in order to do the replacement.
```
Comment (andrewkroh) [inline ecs-update/ecs-update.go:1964]: I think there are few additional attributes that are commonly overridden to modify the indexing behavior. If any of those exist then we would not be able to drop the definition. See

https://github.com/andrewkroh/fydler/blob/ad0a63f364ca24cfe5afc33607b34783946ef689/internal/analysis/useecs/useecs.go#L122-L133

Comment (efd6) [reply ecs-update/ecs-update.go:1964]: Will this need changing in the `fieldsYMLUseExternalECS` func? ISTM yes.

File: ecs-update/ecs-update.go (hunk 3 above)
Comment (efd6) [reply ecs-update/ecs-update.go]: Not yet correctly handled.

File: ecs-update/ecs-update_test.go [hunk 6]
```diff
@@ -380,3 +380,148 @@ func TestReplaceECSFields(t *testing.T) {
 		t.Errorf("unexpected changes found in %s", filepath.Base(fieldsFile.Path()))
 	}
 }
+
+func TestRemoveECSFields(t *testing.T) {
```
Comment (andrewkroh) [inline ecs-update/ecs-update_test.go:384]: Here's an edge case to test/handle.

Given this content:

```
- name: container
  type: group
  fields:
    - name: id
      external: ecs
```

After the `id` is removed, validation of the package will fail with

> field 0.fields: Invalid type. Expected: array, given: null

It looks like the empty parent node needs purged too.

File: ecs-update/ecs-update.go (hunk 5 above)
Comment (andrewkroh) [reply ecs-update/ecs-update.go]: Oh you right. I guess I discovered this problem after adding `-fields-yml-use-ecs` while tinkering with fydler.

File: ecs-update/ecs-update.go [hunk 7]
```diff
@@ -1798,13 +2026,100 @@ func fieldsYMLUseExternalECS(f *ast.File, ff fleetpkg.FieldsFile) (changed bool,
...
+	}
+
+	for _, field := range flatFields {
+		// Does the field exist in ECS?
+		ecsField, _ := ecs.Lookup(field.Name, "")
+		if ecsField == nil {
+			continue
+		}
+
+		if !couldBeECS(field, ecsField) {
```
Comment (andrewkroh) [inline ecs-update/ecs-update.go:2053]: I was testing this and it appears that there are four fields that must be kept no matter what. I'm not sure where this check comes from.

```
Error: building package failed: invalid content found in built zip package: found 4 validation errors:
   1. expected field "data_stream.dataset" with type "constant_keyword" not found in datastream "audit"
   2. expected field "data_stream.namespace" with type "constant_keyword" not found in datastream "audit"
   3. expected field "@timestamp" with type "date" not found in datastream "audit"
   4. expected field "data_stream.type" with type "constant_keyword" not found in datastream "audit"
```

**Comment by andrewkroh** [comment]
I notice on minor commit message issue:

`ecs-update -owner elastic/security-service-integrations -fields-yml-drop-ecs -kibana-version 8.14.0 -drop-import-mappings packages/zoom`

> [zoom] - change to kibana constraint to 8.14.0
> 
> The conditions.kibana.version in the package manifest changed from ^8.12.0 to
> 8.14.0. Modified the field definitions to.

**Review by andrewkroh — APPROVED** [review]
LGTM
